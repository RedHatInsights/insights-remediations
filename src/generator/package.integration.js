'use strict';

const base = require('../test');
const { request } = base;

describe('patchman - package using patch-package', function () {
    test('generates a simple playbook with single patch package update remediation', async () => {
        const data = {
            issues: [{
                id: 'patch-package:rpm-4.14.2-37.el8.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }]
        };

        const res = await request
        .post('/v1/playbook')
        .send(data)
        .expect(200);

        expect(res.text).toMatchSnapshot();
    });

    test('generates a playbook with multiple packages to upgrade', async () => {
        const data = {
            issues: [{
                id: 'patch-package:rpm-4.14.2-37.el8.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }, {
                id: 'patch-package:systemd-239-13.el8_0.5.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }, {
                id: 'patch-package:libstdc++-8.3.1-5.1.el8.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }, {
                id: 'patch-package:qemu-guest-agent-15:4.2.0-34.module+el8.3.0+8829+e7a0a3ea.1.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }, {
                id: 'patch-package:some.my-odd_pkg-1000:11.23.444.5-8.1.el8.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }]
        };

        const res = await request
        .post('/v1/playbook')
        .send(data)
        .expect(200);

        expect(res.text).toMatchSnapshot();
    });

    test('400s on unknown issue', async () => {
        const {id, header} = base.reqId();

        const {body} = await request
        .post('/v1/playbook')
        .set(header)
        .send({
            issues: [{
                id: 'patch-package:unknown-1.2.3',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }]
        })
        .expect(400);

        body.errors.should.eql([{
            id,
            status: 400,
            code: 'UNKNOWN_ISSUE',
            title: 'Unknown issue identifier "patch-package:unknown-1.2.3"'
        }]);
    });
});

// TODO remove after removing package support in "patch-advisory".
describe('patchman - package using patch-advisory', function () {
    test('generates a simple playbook with single patch package update remediation', async () => {
        const data = {
            issues: [{
                id: 'patch-advisory:rpm-4.14.2-37.el8.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }]
        };

        const res = await request
        .post('/v1/playbook')
        .send(data)
        .expect(200);

        expect(res.text).toMatchSnapshot();
    });

    test('generates a playbook with multiple packages to upgrade', async () => {
        const data = {
            issues: [{
                id: 'patch-advisory:rpm-4.14.2-37.el8.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }, {
                id: 'patch-advisory:systemd-239-13.el8_0.5.x86_64',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }]
        };

        const res = await request
        .post('/v1/playbook')
        .send(data)
        .expect(200);

        expect(res.text).toMatchSnapshot();
    });

    test('400s on unknown issue', async () => {
        const {id, header} = base.reqId();

        const {body} = await request
        .post('/v1/playbook')
        .set(header)
        .send({
            issues: [{
                id: 'patch-advisory:unknown-1.2.3',
                systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
            }]
        })
        .expect(400);

        body.errors.should.eql([{
            id,
            status: 400,
            code: 'UNKNOWN_ISSUE',
            title: 'Unknown issue identifier "patch-advisory:unknown-1.2.3"'
        }]);
    });
});
