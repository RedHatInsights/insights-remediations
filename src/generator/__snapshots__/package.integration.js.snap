// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patchman - package using patch-advisory generates a playbook with multiple packages to upgrade 1`] = `
"---
# Red Hat Insights has recommended one or more actions for you, a system administrator, to review and if you
# deem appropriate, deploy on your systems running Red Hat software. Based on the analysis, we have automatically
# generated an Ansible Playbook for you. Please review and test the recommended actions and the Playbook as
# they may contain configuration changes, updates, reboots and/or other changes to your systems. Red Hat is not
# responsible for any adverse outcomes related to these recommendations or Playbooks.

# Upgrade the following packages:
#   - Upgrade packages rpm-4.14.2-37.el8.x86_64
#   - Upgrade packages systemd-239-13.el8_0.5.x86_64
# Identifier: (patch-advisory:rpm-4.14.2-37.el8.x86_64,patch-advisory:systemd-239-13.el8_0.5.x86_64,fix)
# Version: test
- name: update packages
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_issues: \\"rpm-4.14.2-37.el8.x86_64 systemd-239-13.el8_0.5.x86_64\\"
    insights_signature_exclude: \\"/hosts,/vars/insights_issues\\"
  become: true
  tasks:
    - name: check for update
      shell: \\"{{ ansible_facts['pkg_mgr'] }} check-update -q {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      check_mode: no
      register: check_out
      failed_when: check_out.rc != 0 and check_out.rc != 100
      args:
        warn: false

    - when: check_out.rc == 100
      name: upgrade package
      shell: \\"{{ ansible_facts['pkg_mgr'] }} update -d 2 -y {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      args:
        warn: false

    - when: check_out.rc == 100
      name: set reboot fact
      set_fact:
        insights_needs_reboot: True

# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: insights-client
      changed_when: false"
`;

exports[`patchman - package using patch-advisory generates a simple playbook with single patch package update remediation 1`] = `
"---
# Red Hat Insights has recommended one or more actions for you, a system administrator, to review and if you
# deem appropriate, deploy on your systems running Red Hat software. Based on the analysis, we have automatically
# generated an Ansible Playbook for you. Please review and test the recommended actions and the Playbook as
# they may contain configuration changes, updates, reboots and/or other changes to your systems. Red Hat is not
# responsible for any adverse outcomes related to these recommendations or Playbooks.

# Upgrade packages rpm-4.14.2-37.el8.x86_64
# Identifier: (patch-advisory:rpm-4.14.2-37.el8.x86_64,fix)
# Version: test
- name: update packages
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_issues: \\"rpm-4.14.2-37.el8.x86_64\\"
    insights_signature_exclude: \\"/hosts,/vars/insights_issues\\"
  become: true
  tasks:
    - name: check for update
      shell: \\"{{ ansible_facts['pkg_mgr'] }} check-update -q {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      check_mode: no
      register: check_out
      failed_when: check_out.rc != 0 and check_out.rc != 100
      args:
        warn: false

    - when: check_out.rc == 100
      name: upgrade package
      shell: \\"{{ ansible_facts['pkg_mgr'] }} update -d 2 -y {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      args:
        warn: false

    - when: check_out.rc == 100
      name: set reboot fact
      set_fact:
        insights_needs_reboot: True

# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: insights-client
      changed_when: false"
`;

exports[`patchman - package using patch-package generates a playbook with multiple packages to upgrade 1`] = `
"---
# Red Hat Insights has recommended one or more actions for you, a system administrator, to review and if you
# deem appropriate, deploy on your systems running Red Hat software. Based on the analysis, we have automatically
# generated an Ansible Playbook for you. Please review and test the recommended actions and the Playbook as
# they may contain configuration changes, updates, reboots and/or other changes to your systems. Red Hat is not
# responsible for any adverse outcomes related to these recommendations or Playbooks.

# Upgrade the following packages:
#   - Upgrade packages libstdc++-8.3.1-5.1.el8.x86_64
#   - Upgrade packages qemu-guest-agent-15:4.2.0-34.module+el8.3.0+8829+e7a0a3ea.1.x86_64
#   - Upgrade packages rpm-4.14.2-37.el8.x86_64
#   - Upgrade packages some.my-odd_pkg-1000:11.23.444.5-8.1.el8.x86_64
#   - Upgrade packages systemd-239-13.el8_0.5.x86_64
# Identifier: (patch-package:libstdc++-8.3.1-5.1.el8.x86_64,patch-package:qemu-guest-agent-15:4.2.0-34.module+el8.3.0+8829+e7a0a3ea.1.x86_64,patch-package:rpm-4.14.2-37.el8.x86_64,patch-package:some.my-odd_pkg-1000:11.23.444.5-8.1.el8.x86_64,patch-package:systemd-239-13.el8_0.5.x86_64,fix)
# Version: test
- name: update packages
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_issues: \\"libstdc++-8.3.1-5.1.el8.x86_64 qemu-guest-agent-15:4.2.0-34.module+el8.3.0+8829+e7a0a3ea.1.x86_64 rpm-4.14.2-37.el8.x86_64 some.my-odd_pkg-1000:11.23.444.5-8.1.el8.x86_64 systemd-239-13.el8_0.5.x86_64\\"
    insights_signature_exclude: \\"/hosts,/vars/insights_issues\\"
  become: true
  tasks:
    - name: check for update
      shell: \\"{{ ansible_facts['pkg_mgr'] }} check-update -q {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      check_mode: no
      register: check_out
      failed_when: check_out.rc != 0 and check_out.rc != 100
      args:
        warn: false

    - when: check_out.rc == 100
      name: upgrade package
      shell: \\"{{ ansible_facts['pkg_mgr'] }} update -d 2 -y {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      args:
        warn: false

    - when: check_out.rc == 100
      name: set reboot fact
      set_fact:
        insights_needs_reboot: True

# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: insights-client
      changed_when: false"
`;

exports[`patchman - package using patch-package generates a simple playbook with single patch package update remediation 1`] = `
"---
# Red Hat Insights has recommended one or more actions for you, a system administrator, to review and if you
# deem appropriate, deploy on your systems running Red Hat software. Based on the analysis, we have automatically
# generated an Ansible Playbook for you. Please review and test the recommended actions and the Playbook as
# they may contain configuration changes, updates, reboots and/or other changes to your systems. Red Hat is not
# responsible for any adverse outcomes related to these recommendations or Playbooks.

# Upgrade packages rpm-4.14.2-37.el8.x86_64
# Identifier: (patch-package:rpm-4.14.2-37.el8.x86_64,fix)
# Version: test
- name: update packages
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_issues: \\"rpm-4.14.2-37.el8.x86_64\\"
    insights_signature_exclude: \\"/hosts,/vars/insights_issues\\"
  become: true
  tasks:
    - name: check for update
      shell: \\"{{ ansible_facts['pkg_mgr'] }} check-update -q {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      check_mode: no
      register: check_out
      failed_when: check_out.rc != 0 and check_out.rc != 100
      args:
        warn: false

    - when: check_out.rc == 100
      name: upgrade package
      shell: \\"{{ ansible_facts['pkg_mgr'] }} update -d 2 -y {{ insights_issues | regex_search('^[a-z]+-(\\\\\\\\d+:)?([\\\\\\\\d]|\\\\\\\\.)+-[a-z0-9._]+\\\\\\\\.[a-z0-9_]+$') }}\\"
      args:
        warn: false

    - when: check_out.rc == 100
      name: set reboot fact
      set_fact:
        insights_needs_reboot: True

# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  vars:
    insights_signature_exclude: \\"/hosts\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: insights-client
      changed_when: false"
`;
